crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,262 @@
+that all Python type checkers should support,
+The desirable type checker behavior has been standardised to some degree for these specified types,
+but the spec is by no means exhaustive.
+There are more kinds of types in Python than are included in the spec,
+and many kinds of types that are left unspecified must be understood by a type checker
+in order for the type checker to accurately and precisely model Python's runtime semantics.
+
+Some types in Python have a known size.
+Most, however, have an infinite number of possible inhabitants and subtypes.
+For any type `T`, the union of all of the proper subtypes[^2] of `T` is exactly equal to `T`.
+
+## Singleton types
+
+A singleton type is a type for which it is known that there is
+(and can only ever be) exactly one known inhabitant of the type at runtime:
+a set of runtime values with size exactly 1.
+For any singleton type in Python, the type's sole inhabitant will always exist
+at the same memory address for the entire duration of a Python program.
+
+Examples of singleton types in red-knot's model of Python include:
+
+- `types.EllipsisType` (the sole inhabitant is `builtins.Ellipsis`;
+    the constructor always returns the same object):
+
+    ```pycon
+    >>> from types import EllipsisType
+    >>> EllipsisType() is EllipsisType() is ... is Ellipsis
+    True
+    >>> id(EllipsisType()) == id(...)
+    True
+    ```
+
+- `types.NotImplementedType`
+    (in the same way as `types.EllipsisType` and `builtins.Ellipsis`,
+    the sole inhabitant is `builtins.NotImplemented`).
+
+- `None` (which can also be spelled as `Literal[None]`).
+    The sole inhabitant of the type is the runtime constant `None` itself.
+
+- `Literal[True]`: the sole inhabitant of the type is the constant `True`.
+
+- `Literal[False]`: the sole inhabitant of the type is the constant `False`.
+
+- `Literal[E.A]`, `Literal[E.B]` or `Literal[E.C]` for the following enum `E`
+    (the sole inhabitant of `Literal[E.A]` is the enum member `E.A`):
+
+    ```py
+    from enum import Enum, auto
+
+    class E(Enum):
+        A = auto()
+        B = auto()
+        C = auto()
+    ```
+
+- A "literal class type": a type representing a single known class (and none of its possible subclasses).
+
+- A "literal function type": a type that represents a single known function
+    (and excludes all other functions, even if they have the same signature).
+
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All Python singleton types are also sealed types and final types; nearly all are single-value types.
```

Comment:
There's still a reference to "final type" here, which may no longer make sense given later changes.